### PR TITLE
Multi await feature for intentional blocking

### PR DESCRIPTION
**Pull request by danielkec**
There was a request for blocking feature for `Multi` and `Single` 
```java
Multi.interval(20, TimeUnit.MILLISECONDS, Executors.newSingleThreadScheduledExecutor())
        .limit(5)
        .forEach(sum::addAndGet)
        .await();
```
```java
Multi.interval(20, TimeUnit.MILLISECONDS, Executors.newSingleThreadScheduledExecutor())
        .limit(5)
        .reduce(Long::sum)
        .await();
```

I guess this needs some broader discussion, it kind of feels like introducing an antipattern. But the truth is other impls has similar feature and users will be expecting it.

Signed-off-by: Daniel Kec <daniel.kec@oracle.com>

### CODE REVIEW

File: common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    default void await(long timeout, TimeUnit unit) {
+        try {
+            this.toCompletableFuture().get(timeout, unit);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new CompletionException(e);
+        }
+    }
+
+    class MultiCompletableFuture extends CompletableFuture<Void> implements MultiCompletionStage {
```
Comment (akarnokd) [inline common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java:68]: `final`

File: common/reactive/src/main/java/io/helidon/common/reactive/Multi.java [hunk 2]
```diff
@@ -843,9 +843,13 @@
      *
      * @param consumer consumer to be invoked for each item
      */
-    default void forEach(Consumer<? super T> consumer) {
-        FunctionalSubscriber<T> subscriber = new FunctionalSubscriber<>(consumer, null, null, null);
+    default MultiCompletionStage forEach(Consumer<? super T> consumer) {
```
Comment (akarnokd) [inline common/reactive/src/main/java/io/helidon/common/reactive/Multi.java:847]: I'd definitely add cancellation support to this. `AutoCloseable` perhaps.

File: common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+     * @param unit    the time unit of the timeout argument
+     * @throws java.util.concurrent.CancellationException if this future was cancelled
+     * @throws CompletionException                        if this future completed exceptionally,
+     *                                                    was interrupted while waiting or the wait timed out
+     */
+    default void await(long timeout, TimeUnit unit) {
+        try {
+            this.toCompletableFuture().get(timeout, unit);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new CompletionException(e);
```
Comment (akarnokd) [inline common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java:64]: I'd unwrap `ExecutionException` here.

File: common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java (hunk 1 above)
Comment (danielkec) [reply common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java:68]: fixed

File: common/reactive/src/main/java/io/helidon/common/reactive/Multi.java (hunk 2 above)
Comment (danielkec) [reply common/reactive/src/main/java/io/helidon/common/reactive/Multi.java:847]: That's pretty cool! :+1:

File: common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java (hunk 3 above)
Comment (danielkec) [reply common/reactive/src/main/java/io/helidon/common/reactive/MultiCompletionStage.java:64]: thx

**Comment by romain-grecourt** [comment]
Can you rebase against master ? you're missing one of the checks from the new infra.

**Comment by danielkec** [comment]
@tomas-langer had proposed `Single` implementing `CompletionStage` which is seems to be quite nicer, so `Multi.forEach` can return`Single<Void>`, `await` methods are part of single on one place only and its possible to return Single on the apis where completables are expected

**Comment by danielkec** [comment]
Can I ask @romain-grecourt and @tomas-langer for review? @Verdent is blocked by this with #1646 Question is if `Single implementing CompletionStage` is preferable or not, Imho I like it as it makes all apis returning Single more comfortable.

**Comment by romain-grecourt** [comment]
@danielkec Yeah, Let's go ahead with Single implementing CompletionStage, we can leverage it in our APIs where we return a CompletionStage.

File: common/reactive/src/main/java/io/helidon/common/reactive/Multi.java [hunk 4]
```diff
@@ -842,10 +855,23 @@
      * Terminal stage, invokes provided consumer for every item in the stream.
      *
      * @param consumer consumer to be invoked for each item
-     */
-    default void forEach(Consumer<? super T> consumer) {
-        FunctionalSubscriber<T> subscriber = new FunctionalSubscriber<>(consumer, null, null, null);
+     * @return MultiCompletionStage
```
Comment (tomas-langer) [inline common/reactive/src/main/java/io/helidon/common/reactive/Multi.java:858]: Wrong javadoc for return type

Comment (tomas-langer) [reply common/reactive/src/main/java/io/helidon/common/reactive/Multi.java:858]: Also the type is clearly visible from Javadoc already, this should describe behavior, not type

**Comment by danielkec** [comment]
Awaits are now preserved when chaining `CompletionStage` methods:
```java
        testMulti()
                .forEach(sum::addAndGet)
                .whenComplete((aVoid, throwable) -> {})
                .thenRun(() -> {})
                .thenAccept(aVoid -> {})
                .await();
```

**Comment by olotenko** [comment]
This seems like a massive breaking change just to get a wrapper for a timed `CompletableFuture.get`.

`onCancel` implementation is incompatible in its behaviour with superclass. `Multi.forEach` can miss cancellation indefinitely. Dangling `Subscribers` are possible.

**Comment by danielkec** [comment]
@olotenko pls elaborate, is it order of?:
```java
                    subscription.request(Long.MAX_VALUE);
                    single.onCancel(subscription::cancel);
```

**Comment by olotenko** [comment]
Yes, but don't you see the whole design of the change is rickety?

A plain static method for two flavours of `await` would serve the purpose. I really don't see the value of rewriting `CompletableFuture` and changing how all the reactive pipelines everywhere behave.

**Pull request merged**